### PR TITLE
Test run of pipeline to test functionality and provide code feedback.

### PR DESCRIPTION
**Pull request by daniel-e-schmidt**
## Contents
Review deadline:
2024-07-29

Main reviewers: @SannaAb, @pauladisanto and @JPWerling 

### The What
1. Test run the pipeline.
2. Provide feedback on code.

### The Why
1. Verify that the pipeline runs as expected and that the instructions for running it are clear.
2. Find potential issues in the code and give suggestions on areas where it can be approved. These may include things such as functionality, unclear code etc..

### The How
Instructions for running the pipeline are provided on Confluence.

### This [update](https://semver.org/) is:
- [X ] **MAJOR** - when you make incompatible API changes
- [ ] **MINOR** - when you add functionality in a backwards compatible manner
- [ ] **PATCH** - when you make backwards compatible bug fixes or documentation/instructions

## Test Procedure
See Confluence.

### Installation and initiation
See Confluence.

### Tests
Run the pipeline and check the output in folder output/clinic and that it contains:
1. Consensus .fasta files.
2. variant_calling.txt files.
3. A qc_clinic.csv file.

### Expected outcome
That the pipeline runs smoothly and that an email containing exit status = 0 is received.
Check that the files in output/clinic exist and are not empty. For the variant_calling files drug resistance mutations should be found in the end of at least one file.

## Verifications
- [ ] Code reviewed by @SannaAb 
- [ ] Code reviewed by @pauladisanto
- [ ] Code reviewed by @JPWerling

### CODE REVIEW

**Comment by CEJ-R** [comment]
Pipeline worked. Left some comments/feedback on your confluence how-to. Will look at the code next; but the pipeline worked!

File: workflow/scripts/vcf_modification.py
Comment (CEJ-R) [inline workflow/scripts/vcf_modification.py:1]: This is a bit of work. 
- First Id say, you need to break this down into smaller functions. As it stands it's very hard to follow the logic.
- Secondly, as in `qc.py` you'd benefit greatly from one separate "Read file" function, returning a dict. This way you'd not have to add so much logic to handing the dataframes. Simply ask the specific dict for that info and you're done.
- Thirdly, don't nest your loops so much. That can very easily lead to a lot of problems. If you must cycle more than twice, do so once, return the data into a var, and cycle that var separately. That way you can add checks inbetween, reducing the issue and raising readability.
- Fourth(ly?): Move the tables of amino acids, positions, drugs, etc. (all constants) to a config. When these need changing you'd do it there, not fiddle around in the code, which can (and will) break more easily from that.

File: workflow/scripts/wrapper.py [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+inpath = config['input_dir']    # Input path containing barcode folders with fastq files
+outpath = config['output_dir']    # Output path for concatenated fastq files
+sample_sheet = pd.read_csv(config['sample_sheet'])    # Sample sheet following Nanopore convention
+
+# Create output directory if it does not exist (but is defined in the config file)
+if not os.path.exists(outpath):
+    os.makedirs(outpath)
+
+# Check if there is a negative control in the sample sheet
+if not sample_sheet['type'].str.contains('neg', na=False).any():
```
Comment (CEJ-R) [inline workflow/scripts/wrapper.py:22]: As above, what if the user specifies this in caps? One way of doing this simple is to use .lower to force both temporarily to be lower case.

File: workflow/scripts/wrapper.py [hunk 2]
```diff
@@ -0,0 +1,49 @@
+#!/usr/bin/env python3
+
+import pandas as pd
+import glob
+import shutil
+import os
+import yaml
+
+with open('config/wrapper_config.yaml', 'r') as f:
```
Comment (CEJ-R) [inline workflow/scripts/wrapper.py:9]: Id' also put all of the below into a function to better handle any weird business that might arise.

**Review by CEJ-R — CHANGES_REQUESTED** [review]
As it stands I'd say the python parts need a bit of work. I'm not sure how they'd behave in a snakemake setting with my proposed changes, but as python code goes they're quite hard to read. Lift out functions, add docstrings, and move away from handing dataframes. If not here, then in future.

File: README.md [hunk 3]
```diff
@@ -1,2 +1,86 @@
 # hbv_nano
-HBV Nanopore
+
+hbv_nano is a pipeline for analysing whole genome Nanopore long-read data for resistance typing of hepatitis B viruses (HBV).
+
+## Installation
+### 1. Clone the repository and change into it.
+Clone the repository and change the location to it. The location should be: 
+```
+.../hbv_nano
```
[the remark this answers is not in the document]
Comment (daniel-e-schmidt) [reply README.md:9]: Absolutely!

File: workflow/scripts/qc.py [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+## Output the results
+# All results
+df_merged.to_csv(f'{output}/qc.csv', index=False)
+
+# Results relevant to the clinic
+df_clinic = df_merged[['read_id','ref','mapped_reads','mapped_reads_rt','qc_pass']]
+
+os.makedirs(f'{output}/clinic', exist_ok=True)
+
+df_clinic.to_csv(f'{output}/clinic/qc.csv', index=False)
```
[the remark this answers is not in the document]
Comment (JPWerling) [reply workflow/scripts/qc.py:64]: @CEJ-R That it should or should not exist?

[the remark this answers is not in the document]
Comment (CEJ-R) [reply workflow/scripts/qc.py:64]: @JPWerling Always exist! We're not code-savages, are we?

File: README.md [hunk 5]
```diff
@@ -1,2 +1,99 @@
...
+```
+If you want to create the environment in a specific location add the flag `-p <path>`
+```
+mamba env create -f workflow/scripts/environment.yaml -p <path>
+```
+
+### 3. Build Apptainer images for the different software needed to run the pipeline.
+Make sure that you are still in the location ``.../hbv_nano``. To build the images run the script `build_images.sh`:
+```
+./build_images.sh
```
Comment (CEJ-R) [inline README.md:31]: Might be good to put this as `bash ./build_images.sh` to avoid confusion/issues with shebang/permissions.

File: README.md [hunk 6]
```diff
@@ -1,2 +1,99 @@
...
+```
+
+### 4. Prepare input fastq.gz files
+The pipeline takes gzipped `.fastq.gz` files in the format `{sample_01}.fastq.gz`, `{sample_02}.fastq.gz`, ...,  `{sample_**}.fastq.gz` as input.
```
Comment (CEJ-R) [inline README.md:35]: Nitpick: Last could be put as `{sample_n}.fastq.gz` to make it more general.

File: README.md
Comment (CEJ-R) [inline README.md:1]: Much clearer this time around!

File: workflow/scripts/copy_files.py [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+import pandas as pd
+import os
+import yaml
+
+
+def copy_files():
+    '''Copy all files relevant for the clinic to the output folder'''
+
+    # Get output folder using snakemake or, if running script independently, directly from config file
+    try:
```
Comment (CEJ-R) [inline workflow/scripts/copy_files.py:12]: Hmm... I'm wondering if this should be moved to the if __name__ thing. There you could add the if-statement and also feed the function with the input. As it now stands I instinctively looked for `copy_files():` input.

File: workflow/scripts/error_rate.py [hunk 8]
```diff
@@ -0,0 +1,25 @@
+#!/usr/bin/env python3
+
+import pandas as pd
+import yaml
+
+def min_err_rate():
+    try:
```
Comment (CEJ-R) [inline workflow/scripts/error_rate.py:7]: Again maybe move this down. I think that would make me as a reader more easily see what makes the function work.

File: workflow/scripts/qc.py [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+    reads_neg_ctrl_rt = df_merged.loc[df_merged['read_id'].str.contains('neg_ctrl'), 'mapped_reads_rt'].values[0]
+
+    # Calculate the ratio of mapped reads of the negative control to the sample and the rt region
+    df_merged['mapped reads neg_ctrl/mapped reads sample'] = round(reads_neg_ctrl / df_merged['mapped_reads'], 3)
+    df_merged['mapped reads neg_ctrl_rt/mapped reads sample_rt'] = round(reads_neg_ctrl_rt / df_merged['mapped_reads_rt'], 3)
+
+    return df_merged
+
+def qc(df_merged):
+    '''Perform QC checks'''
```
Comment (CEJ-R) [inline workflow/scripts/qc.py:51]: Might want some more info here. QC of what? On what?

File: workflow/scripts/wrapper.py [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+# Create output directory if it does not exist (but is defined in the config file)
+if not os.path.exists(outpath):
+    os.makedirs(outpath)
+
+# Check if there is a negative control in the sample sheet
+if not sample_sheet['type'].str.contains('neg', case=False, na=False).any():
+    raise ValueError("No negative control could be found, please check input data. Aborting script.")
+
+
+def concatenate(sample_sheet, inpath, outpath):
```
Comment (CEJ-R) [inline workflow/scripts/wrapper.py:26]: This would be good if you add some comments in the code. E.g. what the smaller steps do.

**Review by CEJ-R — COMMENTED** [review]
Where you've improved it's much better. Good work there! As expected the places where you've not touched yet still needs some work. As for the new comments, there are just some small stuff.
Overall: Much better!

File: workflow/Snakefile [hunk 11]
```diff
@@ -0,0 +1,262 @@
...
+
+# Preprocess the fastq files
+rule fastp:
+    input:
+        fastq = f'{config["fastq_folder"]}/{{sample}}.fastq.gz'    # get all fastq files
+    output: 
+        fastq = f'{config["output"]}/fastp/{{sample}}_filt.fastq.gz',    # output filtered reads
+        report = f'{config["output"]}/fastp/report/{{sample}}.html'
+    singularity:
+        config["fastp"]
```
Comment (JPWerling) [inline workflow/Snakefile:43]: You mentioned snakemake8 bringing some special functionality running singularities. This looks very standard?

File: workflow/scripts/copy_files.py [hunk 12]
```diff
@@ -0,0 +1,38 @@
...
+import pandas as pd
+import os
+import yaml
+
+try:
+    output = snakemake.params.output
+except NameError:
+    with open('config/config.yaml', 'r') as f:
+        config = yaml.safe_load(f)
+    output = config['output']
```
Comment (JPWerling) [inline workflow/scripts/copy_files.py]: When is `snakemake.params.output` supposed to be filled? Will this not succeed with an empty string for the param?

File: workflow/scripts/copy_files.py [hunk 13]
```diff
@@ -0,0 +1,38 @@
...
+
+
+# Copy all files relevant for the clinic to the output folder
+def copy_files():
+    df = pd.read_csv(f'{output}/samtools/minimum_error_rates.csv')
```
Comment (JPWerling) [inline workflow/scripts/copy_files.py]: Consider less "bloated", more vanilla way of parsing csv:
```
import csv
with open('names.csv') as csvfile:
    reader = csv.DictReader(csvfile)
    for row in reader:
        print(row['first_name'], row['last_name'])
```

File: workflow/scripts/copy_files.py [hunk 14]
```diff
@@ -0,0 +1,38 @@
...
+        consensus_in = f'{output}/consensus/medaka/{read_id}.{ref}/consensus.fasta'
+        consensus_out = f'{output}/clinic/{read_id}.{ref}.fa'
+
+        os.makedirs(f'{output}/clinic', exist_ok=True)
+
+        # Change headers of consensus files and write to output folder
+        with open(consensus_in, 'r') as in_fasta, open(consensus_out, 'w') as out_fa:
+            for line in in_fasta:
+                if line.startswith('>'):
+                    out_fa.write(f'>{read_id}.{ref}\n')
```
Comment (JPWerling) [inline workflow/scripts/copy_files.py:34]: `print(f'>{read_id}.{ref}', file=out_fa)` is an alternative if you dont want to add newlines all the time.

File: workflow/scripts/environment.yaml [hunk 15]
```diff
@@ -0,0 +1,8 @@
+name: hbv_nano
+channels:
+  - bioconda
+  - conda-forge
+dependencies:
+  - snakemake-minimal
+  - biopython
+  - pandas
```
Comment (JPWerling) [inline workflow/scripts/environment.yaml:8]:
```
name: hbv_nano
channels:
  - bioconda
  - conda-forge
dependencies:
  - python=3.12
  - snakemake-minimal=8.18.0
  - pip=24.2
  - pip:
    - biopython==1.84
    - pandas==2.2.2
```

File: workflow/scripts/error_rate.py [hunk 16]
```diff
@@ -0,0 +1,21 @@
...
+        config = yaml.safe_load(f)
+    output = config['output']
+
+# Calculate the minimum error rate for each read_id
+input_file = f'{output}/samtools/error_rates.csv'
+output_file = f'{output}/samtools/minimum_error_rates.csv'
+
+df = pd.read_csv(input_file, header=None)    # import error rates
+df.columns = ["read_id", "ref", "error_rate"]
+df = df.loc[df.groupby("read_id")["error_rate"].idxmin()].reset_index(drop=True)   # get the sample with the lowest error rate for each read_id
```
Comment (JPWerling) [inline workflow/scripts/error_rate.py]: This is impossible to understand for anyone not well-read on all things pandas

**Review by JPWerling — CHANGES_REQUESTED** [review]
Few points i think could be improved.

I made some changes of my own in the git branch `golden` where I try and implement some stricter versioning of dependencies. You could look there and implement the changes yourself here or leave the branch as is, but the versions need to be set more specifically for both python env and singu images
